File: python/leetcode/069.py
class Solution(object):
    def mySqrt(self, x):
        if x <= 1:
            return x
        l, r = 0, x
        while l <= r:
            m = l + (r - l) / 2
            if m * m == x:
                return m
            elif m * m > x:
                r = m - 1
            else:
                l = m + 1
        return r

File: cpp/leetcode/069.cpp
int mySqrt(int x) {
    long r = x;
    while (r * r > x) {
        r = (r + x / r) / 2;
    }
    return r;
}

class Solution {
public:
    int mySqrt(int x) {
        long long l = 0, r = x;
        while (l <= r) {
            long long m = (r - l) / 2 + l;
            if (m * m > x) {
                r = m - 1;
            } else if (m * m == x) {
                return m;
            } else {
                l = m + 1;
            }
        }
        return r;
    }
};